### PR TITLE
Directive ends on complete unindent

### PR DESCRIPTION
**Pull request by linawolf**
Expecially in code blocks the first line of the directive content can have more indentation then the following ones. Overall this is true for all directives.

The general rule is that a directive ends when the indentation level of the directive declaration is found or when the file ends.

### CODE REVIEW

File: packages/guides-restructured-text/src/RestructuredText/Parser/Productions/DirectiveRule.php [hunk 1]
```diff
@@ -184,20 +184,19 @@ private function parseDirectiveOption(string $line): DirectiveOption
     private function collectDirectiveContents(LinesIterator $documentIterator): Buffer
     {
         $buffer = new Buffer();
-        $indenting = 1;
-        while (LinesIterator::isBlockLine($documentIterator->getNextLine(), $indenting)) {
+        $minIndenting = 1000;
```
Comment (jaapio) [inline packages/guides-restructured-text/src/RestructuredText/Parser/Productions/DirectiveRule.php:187]: Could you please extract the default value to a constant, so we can add docblocks above it to explain what it is? Right now it feels like a random number.

File: packages/guides-restructured-text/src/RestructuredText/Parser/Productions/DirectiveRule.php [hunk 2]
```diff
@@ -184,20 +184,19 @@ private function parseDirectiveOption(string $line): DirectiveOption
...
+                $minIndenting = min($minIndenting, $indenting);
             }
-
-            if ($line !== '') {
-                $line = substr($documentIterator->current(), $indenting);
-            }
-
             $buffer->push($line);
         }
+        if ($minIndenting < 1000) {
```
Comment (jaapio) [inline packages/guides-restructured-text/src/RestructuredText/Parser/Productions/DirectiveRule.php:197]: Is this not equal to always?

File: packages/guides-restructured-text/src/RestructuredText/Parser/Buffer.php [hunk 3]
```diff
@@ -84,4 +84,13 @@ public function clear(): void
     {
         $this->lines = [];
     }
+
+    public function unIndent(int $indentation): void
```
Comment (jaapio) [inline packages/guides-restructured-text/src/RestructuredText/Parser/Buffer.php:88]: I get what you are doing here, but it feels like something that can be forgotten when using this class. Should we add this in the `getLinesString` method and `getLines`? with the default of indenting 1?

**Review by jaapio — CHANGES_REQUESTED** [review]
A few suggestions to improve the code. 

And it would be nice if we had more unit tests. The functional tests are beneficial, but they are much slower. So if possible it would be nice to add some tests that cover the edge cases of the `DirectiveRule` You can have a look at the other Rule tests for inspiration.

File: packages/guides-restructured-text/src/RestructuredText/Parser/Productions/DirectiveRule.php (hunk 1 above)
Comment (linawolf) [reply packages/guides-restructured-text/src/RestructuredText/Parser/Productions/DirectiveRule.php:187]: It is a random number it just needs to be very high...

File: packages/guides-restructured-text/src/RestructuredText/Parser/Buffer.php (hunk 3 above)
Comment (linawolf) [reply packages/guides-restructured-text/src/RestructuredText/Parser/Buffer.php:88]: Mmh, I don't really get what you suggest..

File: packages/guides-restructured-text/src/RestructuredText/Parser/Productions/DirectiveRule.php (hunk 1 above)
Comment (linawolf) [reply packages/guides-restructured-text/src/RestructuredText/Parser/Productions/DirectiveRule.php:187]: Mmh, maybe with PHP_INT_MAX ?

File: packages/guides-restructured-text/src/RestructuredText/Parser/Buffer.php (hunk 3 above)
Comment (jaapio) [reply packages/guides-restructured-text/src/RestructuredText/Parser/Buffer.php:88]: I would like to see the effect of this method, be part of the `getLinesString` method and `getLines`

eg. 

```
getLines(int $indentation = 1)
{
   return array_map(function($line) {
            if (strlen($line) >= $indentation) {
                return substr($line, $indentation);
            }
            
            return $line;
      }
   }
}
```
This way we can not forget to remove the indentation, we have multiple places where we use the buffer and strip the first part of the indenting before we start another cycle of parsing the content of a block. Those can be replaced this way.

Comment (linawolf) [reply packages/guides-restructured-text/src/RestructuredText/Parser/Buffer.php:88]: In the very place where I used the unIndent function getLines is not used, just the buffer returned. So we would have to return the information about the depth of the indentation and the buffer only to apply unindent lateron?

Comment (jaapio) [reply packages/guides-restructured-text/src/RestructuredText/Parser/Buffer.php:88]: I have created a list of possible locations to dig into: 

- This method removes the indenting like you do. Based on the first line of a block.
 `\phpDocumentor\Guides\RestructuredText\Parser\Productions\BlockQuoteRule::normalizeLines`
- Here I do something simular for lists
  `packages/guides-restructured-text/src/RestructuredText/Parser/Productions/ListRule.php:107`
- `packages/guides-restructured-text/src/RestructuredText/Parser/Productions/EnumeratedListRule.php:109`
- Definition lists: `packages/guides-restructured-text/src/RestructuredText/Parser/Productions/DefinitionListRule.php:99`

Comment (jaapio) [reply packages/guides-restructured-text/src/RestructuredText/Parser/Buffer.php:88]: Yes, that would work for me. I will turn this into an issue.

**Review by jaapio — APPROVED** [review]
